### DISCUSSION TITLE
Data Migration from ELSA 2 to ELSA 3: Real World experience?

### DISCUSSION THREAD

**Discussion by r-kernberger** [elsa-workflows/elsa-core#6223] [Q&A]
After migrating our ELSA 2 Code to ELSA 3 there is one topic unsolved: Data Migration.

Did anyone migrate existing workflow instances from ELSA 2 to ELSA 3 on database level already and has gone in production?

After comparing the data structure of v2 and v3 I've come to the conclusion, that they are too different. 
Especially the "data" field in the workflowinstances table.

So, my approach will be to set up an ELSA 3 System in parallel to our ELSA 2 System. 

New workflow instances will only be created in ELSA 3. 
Existing workflow instances in ELSA 2 will stay there until they are finished. 
After that: will turn off ELSA 2

Any other approaches / experience? Would be happy to hear.

**Comment by sfmskywalker** [comment]
Yes—your DBA/CI process can create the MongoDB collections and indexes before Elsa starts, provided it uses the collection names and index definitions expected by the Elsa MongoDB persistence module.

The current provider resolves fixed collection names such as `workflow_instances`, `workflow_execution_logs`, `bookmarks`, and `bookmark_queue_items`, and its startup hosted services call `CreateManyAsync` for the required indexes. Creating those same collections/indexes in CI/CD is compatible; keep the definitions aligned when upgrading Elsa. The relevant implementation is in `Elsa.Persistence.MongoDB`:

- `Modules/Management/WorkflowManagementPersistenceFeature.cs`
- `Modules/Runtime/WorkflowRuntimePersistenceFeature.cs`
- `Modules/Management/CreateIndices.cs`
- `Modules/Runtime/CreateIndices.cs`

Elsa does not configure TTL indexes itself. A DBA-managed TTL index is technically possible, but it deletes documents independently of Elsa, so use it only with a retention policy that cannot remove data an active or suspended workflow still needs. In particular, do **not** TTL-delete `workflow_instances`, `bookmarks`, or `bookmark_queue_items` for active/suspended work. TTL retention is much safer for terminal workflow instances and execution/activity logs, with a policy that preserves any records needed for your operational and audit requirements.

There is no reliable fixed read/write estimate from “2,600 workflows per day”: it depends on workflow shape, persistence of bookmarks, triggers, variables, activity execution logs, retries, and Studio/API queries. The practical next step is to load-test representative workflows against MongoDB monitoring and size the indexes/retention policy from those measurements.
